### PR TITLE
Connection pool for Contract Dao

### PR DESCRIPTION
**Pull request by akshayshirahatti-da**
Switch ContractDao to use a HikariCP connection pool based on performance improvements suggested by some naive gatling testing.
https://docs.google.com/spreadsheets/d/1HNqzOlMJX0X963DSGDZAYmk3JbWJHwvCbC4TtkUP9iQ/edit?usp=sharing

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by cocreature — APPROVED** [review]
Thanks! Loos great

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Connection.scala [hunk 1]
```diff
@@ -16,3 +23,41 @@ object Connection {
     Transactor
       .fromDriverManager[IO](jdbcDriver, jdbcUrl, username, password)(IO.ioConcurrentEffect(cs), cs)
 }
+
+object ConnectionPool {
+
+  // below values are hardcoded for now, can be passed as props later.
+  final val MinIdle = 8
+  final val IdleTimeout = 10000
+  final val PoolSize = 10
```
Comment (cocreature) [inline ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Connection.scala:32]: Definitely in favor of exposing these via a cli flag. If nothing else, it makes it easier to experiment with different parameters.

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Connection.scala [hunk 2]
```diff
@@ -3,16 +3,79 @@
...
+      .fromDriverManager[IO](cfg.driver, cfg.url, cfg.user, cfg.password)(
+        IO.ioConcurrentEffect(cs),
+        cs,
+      )
+}
+
+object ConnectionPool {
+
+  /*
+  TODO:
```
Comment (akshayshirahatti-da) [inline ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Connection.scala:33]: Added TODO's to refactor duplicate code and add configurable props, which I'll get to in a follow-up PR

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Connection.scala [hunk 3]
```diff
@@ -3,16 +3,79 @@
...
+    val c = new HikariConfig()
+    c.setJdbcUrl(cfg.url)
+    c.setUsername(cfg.user)
+    c.setPassword(cfg.password)
+    c.setMinimumIdle(MinIdle)
+    c.setPoolName("json-api-jdbc-pool")
+    c.setConnectionTimeout(ConnectionTimeout)
+    c.setMaximumPoolSize(poolSize)
+    c.setIdleTimeout(IdleTimeout) // ms, minimum according to log, defaults to 600s
+    c.setMaximumPoolSize(poolSize)
```
Comment (cocreature) [inline ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Connection.scala:78]:
```suggestion
```
setting it twice is presumably harmless but a bit confusing.

**Review by cocreature — APPROVED** [review]
Nice work! A few minor comments

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala [hunk 4]
```diff
@@ -28,9 +39,31 @@ class ContractDao private (xa: Connection.T)(implicit val jdbcDriver: SupportedJ
...
+      _ <- Try { ds.close() }
+    } yield ()
+  }
+
+  @throws[IOException]
+  override def close(): Unit = {
+    shutdown().fold(
+      {
+        case e: IOException => throw e
+        case e => throw new IOException(e)
```
[the remark this answers is not in the document]
Comment (akshayshirahatti-da) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala:58]: it was to purely preserve the Closeable interface contract (https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/share/classes/java/io/Closeable.java) which throws an IOException.

File: ledger-api/testing-utils/src/main/scala/com/digitalasset/ledger/api/testing/utils/AkkaBeforeAndAfterAll.scala [hunk 5]
```diff
@@ -34,8 +34,19 @@ trait AkkaBeforeAndAfterAll extends BeforeAndAfterAll {
       )
     )
 
-  protected implicit lazy val system: ActorSystem =
-    ActorSystem(actorSystemName, defaultExecutionContext = Some(executionContext))
+  protected implicit lazy val system: ActorSystem = {
+    val customConf = ConfigFactory
+      .parseString("""
+        akka.http.server.request-timeout = 60 s
```
[the remark this answers is not in the document]
Comment (akshayshirahatti-da) [reply ledger-api/testing-utils/src/main/scala/com/digitalasset/ledger/api/testing/utils/AkkaBeforeAndAfterAll.scala]: this change is no longer needed and should be reverted now as we know that it was the connection timeout causing the issue thanks to you :)

**Comment by akshayshirahatti-da** [comment]
work done as part of [#10155](https://github.com/digital-asset/daml/issues/10155)

**Pull request merged**